tests/Feature/SharingSpecialTest.php

```diff
@@ -0,0 +1,493 @@
+			true,
+			$this->photoID5, [ // albums 1-3 and 5 are accessible, photo 5 is alphabetically first
+				$this->generateExpectedPhotoJson(self::SAMPLE_FILE_SUNSET_IMAGE, $this->photoID5, $this->albumID5, ['title' => 'Abenddämmerung']),
+				$this->generateExpectedPhotoJson(self::SAMPLE_FILE_MONGOLIA_IMAGE, $this->photoID3, $this->albumID3),
+				$this->generateExpectedPhotoJson(self::SAMPLE_FILE_NIGHT_IMAGE, $this->photoID1, $this->albumID1),
+				$this->generateExpectedPhotoJson(self::SAMPLE_FILE_TRAIN_IMAGE, $this->photoID2, $this->albumID2),
+			]
+		));
+		foreach ([$this->albumID4, $this->photoID4, $this->albumID6, $this->photoID6] as $id) {
+			$responseForRecent->assertJsonMissing(['id' => $id]);
+		}
+
+		// 3. Check tree
+
+		$responseForTree = $this->root_album_tests->getTree();
+		$responseForTree->assertJson($this->generateExpectedTreeJson([
+			$this->generateExpectedAlbumJson(
+				$this->albumID1,
+				self::ALBUM_TITLE_1,
+				null,
+				$this->photoID5,
+				['albums' => [
+					$this->generateExpectedAlbumJson($this->albumID2, self::ALBUM_TITLE_2, $this->albumID1, $this->photoID2),
+					$this->generateExpectedAlbumJson($this->albumID3, self::ALBUM_TITLE_3, $this->albumID1, $this->photoID3),
+					// album 4 has been unlocked simultaneously with password 1, but is hidden and hence not part of the tree
+					$this->generateExpectedAlbumJson($this->albumID5, self::ALBUM_TITLE_5, $this->albumID1, $this->photoID5),
+					// album 6 has been unlocked simultaneously with password 2, but is hidden and hence not part of the tree
+				]]
+			),
+		]));
+		foreach ([
+			$this->photoID1,
+			$this->albumID4,
+			$this->photoID4,
+			$this->albumID6,
+			$this->photoID6,
+		] as $id) {
+			$responseForTree->assertJsonMissing(['id' => $id]);
+		}
+
+		// 4. Check album/photo access
+
+		foreach ([
+			[$this->albumID1, self::ALBUM_TITLE_1, null, $this->photoID5], // photo 5 is alphabetically first of all child photos
+			[$this->albumID2, self::ALBUM_TITLE_2, $this->albumID1, $this->photoID2],
+			[$this->albumID3, self::ALBUM_TITLE_3, $this->albumID1, $this->photoID3],
+			[$this->albumID4, self::ALBUM_TITLE_4, $this->albumID1, $this->photoID4], // album 4 has been unlocked simultaneously with password 1 and hence is directly accessible, although it is hidden
+			[$this->albumID5, self::ALBUM_TITLE_5, $this->albumID1, $this->photoID5],
+			[$this->albumID6, self::ALBUM_TITLE_6, $this->albumID1, $this->photoID6], // album 6 has been unlocked simultaneously with password 2 and hence is directly accessible, although it is hidden
+		] as $albumInfo) {
+			$responseForAlbum = $this->albums_tests->get($albumInfo[0]);
+			$responseForAlbum->assertJson($this->generateExpectedAlbumJson(
+				$albumInfo[0], $albumInfo[1], $albumInfo[2], $albumInfo[3]
+			));
+		}
+		foreach ([
+			$this->photoID1,
+			$this->photoID2,
+			$this->photoID3,
+			$this->photoID4,
+			$this->photoID5,
+			$this->photoID5,
```

Comment:
A typo: this should be `photoID6`

Reply:
Well spotted.

I wonder how we include these findings after the PR has already been merged. I will probably create a bug-fix PR after you have finished your review based on your findings.

Reply:
Personally, I would include a fix in the next (possibly unrelated) PR, simply because this is not fixing any critical bug.

Anyway, I'm done with my quick walkthrough of this PR.